Add support for packaging VTK SDK as a wheel

AlexyPellegrini opened the pull request:
This is the first batch of code to build the sdk wheel!

The URL of the SDK is composed from the target information, and no hash is checked for now, it was easier to test since my python versions mismatch between windows and linux :) 

 I'm not sure how we are supposed to propagate the version using the dynamic metadata, for now it seems to always be v0.1.

Also I see that on "cmake-python-distributions" [there is a script](https://github.com/scikit-build/cmake-python-distributions/blob/560c40d038149556651fe8b01e66093c01aceb30/scripts/update_cmake_version.py) to update the version in different files, is this something that should be done for this repo too?

jcfr commented:
> we are supposed to propagate the version using the dynamic metadata, for now it seems to always be v0.1

There are two aspects to consider:
* The version of the SDK to download will be hardcoded in the CMakeLists.txt
* The version version associated with the wheel will be "dynamically" set based on the git tag

AlexyPellegrini commented:
Hmm it seems that pylint triggers on VTK files, should we disable it since we won't have any python code?

AlexyPellegrini commented:
> > we are supposed to propagate the version using the dynamic metadata, for now it seems to always be v0.1
> 
> There are two aspects to consider:
> 
> * The version of the SDK to download will be hardcoded in the CMakeLists.txt
> * The version version associated with the wheel will be "dynamically" set based on the git tag

Ok, so we will tag later I guess. For now the version I hardcoded is 9.2.5. If we hardcode the version maybe I can bring back the hashes of the archives. Can we use the python 3.8 archive version for any  newerpython version, or should each version match exactly ?

jcfr commented on cmake/vtk-sdk-urls.cmake:
```diff
@@ -0,0 +1,13 @@
+# VTK SDK is currently stored as a tar.xz archive on a Kitware hosted server
+set(python_cp "cp${Python_VERSION_MAJOR}${Python_VERSION_MINOR}")
+
+# enable any linux to download the right tar.xz
+if(${CMAKE_SYSTEM_NAME} STREQUAL "Linux")
+  set(VTK_SDK_BINARY_URL
+      "https://vtk.org/files/wheel-sdks/vtk-wheel-sdk-${VTK_VERSION}-${python_cp}-${python_cp}-manylinux_2_17_x86_64.manylinux2014_x86_64.tar.xz"
+  )
+else()
+  set(VTK_SDK_BINARY_URL
+      "https://vtk.org/files/wheel-sdks/vtk-wheel-sdk-${VTK_VERSION}-${python_cp}-${Python_SOABI}.tar.xz"
+  )
+endif()
```
```suggestion
set(VTK_SDK_SOABI ${Python_SOABI})
if(LINUX AND CMAKE_SIZEOF_VOID_P EQUAL 8)
  set(VTK_SDK_SOABI "manylinux_2_17_x86_64.manylinux2014_x86_64")
endif()
set(VTK_SDK_BINARY_URL
      "https://vtk.org/files/wheel-sdks/vtk-wheel-sdk-${VTK_VERSION}-${python_cp}-${VTK_SDK_SOABI}.tar.xz"
  )
```

jcfr replied:
`LINUX AND CMAKE_SIZEOF_VOID_P EQUAL 8` expression should be augmented to check if we are on a x86_64 platform (and not arm)

jcfr commented:
> a script [...] to update the version in different files, is this something that should be done for this repo too?

Definitely, we should look into adding a script called `update_vtk_sdk_version.py` that would be similar to `update_cmake_version.py` and also have a logic similar to the following:


```python
vtk_version = "9.3.0"

for py_abi_tag in ["cp38-cp38", "cp39-cp39", "cp310-cp310", "cp311-cp311", "cp312-cp312"]:
    for platform_tag in ["macosx_10_10_x86_64", "macosx_11_0_arm64", "manylinux_2_17_x86_64.manylinux2014_x86_64", "win_amd64"]:
        url = f"https://vtk.org/files/wheel-sdks/vtk-wheel-sdk-{vtk_version}-{py_abi_tag}-{platform_tag}.tar.xz"

        # Download files
        # -> Adapt from https://github.com/girder/slicer_package_manager/blob/342b48d7b6fdf407731513c30f5dc147eca30d5b/tests/__init__.py#L324-L334

        # Compute checksum 
        # -> See "computeFileChecksum" at https://github.com/girder/slicer_package_manager/blob/342b48d7b6fdf407731513c30f5dc147eca30d5b/tests/__init__.py#L232-L254
```

jcfr commented:
> Hmm it seems that pylint triggers on VTK files, should we disable it since we won't have any python code?

I ended up addressing the "root" cause by installing the content of the archive in a dedicated directory without relying on the automatic discovery of wheel packages.

AlexyPellegrini commented:
Thank you for making the required changes!

It seems that the mac runners fails because the platform tag (macosx_12_0...) is not the same as the one on the sdk repo. We probably need to the same thing as we do for linux. 
There is an additional "cpN" in the windows url, I think that's because on WIndows the Python SOABI already contains "cpN" at the beginning.

 I'm working on the script to generate the cmake urls file so we will get rid of the problem by changing the logic I think (since the URLs will be hardcoded)

jcfr commented:
Before moving forward with the integration, we should add a test checking that a simple project can be built against the generated wheel.

Add files like the following:

```
tests/test_find_package.py
tests/packages/find_package/CMakeLists.txt
tests/packages/find_package/pyproject.toml
tests/packages/find_package/src/vtk_simple/__init__.py
```

Consider borrowing some of the testing logic found here:
* https://github.com/scikit-build/scikit-build-core/tree/main/tests
* https://github.com/scikit-build/scikit-build-core/blob/main/tests/conftest.py

In this simple test, the `CMakeLists.txt` will be similar to this:

```
cmake_minimum_required(VERSION 3.15...3.26)

project(
  ${SKBUILD_PROJECT_NAME}
  LANGUAGES C
  VERSION ${SKBUILD_PROJECT_VERSION})

find_package(Python COMPONENTS Interpreter Development.Module)

set(expected_version "9.2.5")
find_package(VTK ${expected_version} REQUIRED)
```

AlexyPellegrini commented:
I'm not sure how I am supposed to fix this error:
tests/test_find_package.py:10: error: Cannot find implementation or library stub for module named "virtualenv"  [import-not-found]
I tried to add virtualenv as a test dependency, but it does not seem to fix the issue.

AlexyPellegrini commented:
> I'm not sure how I am supposed to fix this error: tests/test_find_package.py:10: error: Cannot find implementation or library stub for module named "virtualenv" [import-not-found] I tried to add virtualenv as a test dependency, but it does not seem to fix the issue.

It seems that virtualenv has no type information so I added an exclusion in the code.
I also added it in the tests' requirements and the CI `additional_dependencies`

I'm not sure I use the best approach, but it seems to work! At least on Mac and Windows. There is an issue on Ubuntu because it can't find OpenGL

AlexyPellegrini commented:
Ready to merge!

AlexyPellegrini commented:
@jcfr Can you please do a quick review for this PR so we can move forward.

I removed hash check because hash text files are not generated automatically, so they are not available for latest versions of VTK. Let's address this issue in another PR.

AlexyPellegrini commented:
FYI, I bumped VTK to 9.5.0 and Python to 3.10+ from 3.8+ @finetjul @Thibault-Pelletier

jcfr commented:
Notes that I re-inroduced the use of checksums. This is critical to ensure integrity.

After locally downloading the wheels .. the following is helpful:

```
sha256sum vtk-wheel-sdk-9.5.0-cp3* | awk '{
  hash=$1; file=$2
  sub(/^.*\//, "", file)                       # strip any path
  sub(/^vtk-wheel-sdk-/, "", file)             # drop prefix
  sub(/\.tar\.xz$/, "", file)                  # drop suffix
  n=split(file, a, "-"); py=a[2]               # cp tag (e.g., cp38, cp39)
  if (NR>1 && py!=lastpy) print ""             # blank line between cp groups
  lastpy=py
  printf "set(sha256_%s\n    \"%s\")\n", file, hash
}'
```

jcfr commented:
Indeed, once the corresponding capabilities  are in upstream VTK, there would be no need to publish the vtk-sdk wheels from this project. 

I was initially thinking we could publish also publish a custom scikit build backend along side the SDK but this should instead be done in a dedicated GitHub project. 

Indeed, the goal being to simplify how project like the following are externally built by leveraging `VTKExternalModule` (a Slicer agnostic project):

https://github.com/Kitware/LookingGlassVTKModule
